### PR TITLE
Unblock WaitForStartCaptureRequest when stop is requested

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
Unblcok `CloudCollectorStartStopCaptureRequestWaiter::WaitForStartCaptureRequest`
when stop capture is requested.

For the CloudCollector, it is possible that StopCapture is called while
we are waiting for the start capture request, for instance, the GuestOrc
connection failed at the very beginning.
In this case, we unblock `WaitForStartCaptureRequest` and terminate the
capture immediately.

Bug: http://b/220888294

### CODE REVIEW

File: src/LinuxCaptureService/LinuxCaptureServiceBase.cpp [hunk 1]
```diff
@@ -239,8 +239,15 @@ CaptureServiceBase::CaptureInitializationResult LinuxCaptureServiceBase::DoCaptu
...
 
-  const CaptureOptions& capture_options =
-      start_stop_capture_request_waiter->WaitForStartCaptureRequest();
+  auto capture_options_or_error = start_stop_capture_request_waiter->WaitForStartCaptureRequest();
+  if (capture_options_or_error.has_error()) {
+    ORBIT_ERROR("An error occurred while waiting for start: %s",
+                capture_options_or_error.error().message());
+    client_capture_event_collector->StopAndWait();
+    TerminateCapture();
+    return CaptureInitializationResult::kSuccess;
```
Comment (beckerhe) [inline src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:248]: It feels wrong to return kSuccess in this case. Is there no better alternative? If not, could we add an additional enum value for this error case?

File: src/WindowsCaptureService/WindowsCaptureService.cpp [hunk 2]
```diff
@@ -41,8 +41,9 @@ grpc::Status WindowsCaptureService::Capture(
               "Cannot start capture because another capture is already in progress"};
   }
 
-  const CaptureOptions& capture_options =
-      start_stop_capture_request_waiter->WaitForStartCaptureRequest();
+  auto capture_options_or_error = start_stop_capture_request_waiter->WaitForStartCaptureRequest();
+  ORBIT_CHECK(capture_options_or_error.has_value());
```
Comment (beckerhe) [inline src/WindowsCaptureService/WindowsCaptureService.cpp:45]: Is it possible to handle the error case instead of checking?

File: src/LinuxCaptureService/LinuxCaptureServiceBase.cpp (hunk 1 above)
Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp]: I added the `kFailure` for this case. And both the `LinuxCaptureService::Capture` and `WindowsCaptureService::Capture` will return the grpc status `INTERNAL` when getting `CaptureInitializationResult::kFailure`.

**Comment by dpallotti** [comment]
I have a higher level question first: if we receive a stop, do we already know that there won't be a start afterwards? In other words, does the cloud collector only live for a single capture?

I'm trying to understand why the stop while waiting for start cannot simply be ignored.

File: src/LinuxCaptureService/LinuxCaptureServiceBase.cpp (hunk 1 above)
Comment (dpallotti) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp]: I'm wondering if `kAlreadyInProgress` isn't already also a failure. In that case `kFailure` would be too generic. I don't think long names are a problem. If we need more detail, I would even go for `kFailureWhileWaitingForStart` as it's grammatically more correct.

**Comment by vickyliu-go4it** [comment]
> 

The CloudCollector only lives for a single capture. It exits right after finishing the capturing and uploading. When a new capture is requested, the GuestOrc will launch a new CloudCollector.

This can be ignore for the real case as the CloudCollector terminate the main executor and exits in that case.

But it causes some issue for the test while I was using a fake executor: the fake executor does not do anything when we call Stop.

**Comment by beckerhe** [comment]
> I have a higher level question first: if we receive a stop, do we already know that there won't be a start afterwards? In other words, does the cloud collector only live for a single capture?
> 
> I'm trying to understand why the stop while waiting for start cannot simply be ignored.

What Vicky said. But you also bring up a good point. Unblocking the Waiter is not really the same as stopping. Maybe we should introduce a separate method for that. Maybe `AbortCapture()` to make it clearer what is happening here. What do you both think?

File: src/LinuxCaptureService/LinuxCaptureServiceBase.cpp [hunk 3]
```diff
@@ -239,8 +239,15 @@ CaptureServiceBase::CaptureInitializationResult LinuxCaptureServiceBase::DoCaptu
...
 
-  const CaptureOptions& capture_options =
-      start_stop_capture_request_waiter->WaitForStartCaptureRequest();
+  auto capture_options_or_error = start_stop_capture_request_waiter->WaitForStartCaptureRequest();
+  if (capture_options_or_error.has_error()) {
+    ORBIT_ERROR("An error occurred while waiting for start: %s",
+                capture_options_or_error.error().message());
+    client_capture_event_collector->StopAndWait();
+    TerminateCapture();
+    return CaptureInitializationResult::kFailureWhileWaitingForStart;
```
Comment (dpallotti) [inline src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:248]: I think there is some confusion introduced here. `CaptureInitializationResult` is supposed to be returned by `InitializeCapture`, while here we are past that.

This is visible by comparing the Linux and the Windows version. 

The Windows version has an `ORBIT_UNREACHABLE();` for `InitializeCapture` returning `kFailureWhileWaitingForStart`. Then it calls `WaitForStartCaptureRequest` and if this returns an error, it returns `StatusCode::INTERNAL`.

The Linux version does something quite different. It doesn't consider whether `InitializeCapture` can return `kFailureWhileWaitingForStart` or not (it can't, right?). Then, if `WaitForStartCaptureRequest` returns an error, it returns `kFailureWhileWaitingForStart`, which is then conditioning returning `StatusCode::INTERNAL` in the caller. 

I think we need to bring the two cases close together.

The problem is that `CaptureInitializationResult` was meant to be the result of `InitializeCapture`, but now `LinuxCaptureServiceBase::DoCapture` is adding a new case to it (`kFailureWhileWaitingForStart`) which is not related to `InitializeCapture` itself.

Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:248]: +2 We also had a similar discussion on this yesterday. I thought about moving `InitializeCapture` out from `DoCapture`.

I will see whether there is a better way to make this change.

Comment (dpallotti) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:248]: Maybe the problem comes from the fact that `DoCapture` "reuses" `CaptureInitializationResult` as the return type. Could it just return `grpc::Status`? I guess the thing was that we wanted to keep `LinuxCaptureServiceBase` "grpc-free"? But maybe it's fine as it's just a status. Otherwise we might need a new "result" type for `DoCapture`/

Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:248]: How about I move both `InitializeCapture` and `WaitForStart` from `LinuxCaptureServiceBase::DoCapture`,
and change both Windows/Linux capture services' `Capture` rpc to have the following workflow
https://screenshot.googleplex.com/C6ifwiNe3pTePyN

Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:248]: > Maybe the problem comes from the fact that DoCapture "reuses" `CaptureInitializationResult` as the return type. Could it just return `grpc::Status`?

This also sounds good to me. Thanks for the suggestion!

File: src/CaptureServiceBase/include/CaptureServiceBase/CaptureServiceBase.h [hunk 4]
```diff
@@ -33,11 +33,11 @@ class CaptureServiceBase {
     kGuestOrcConnectionFailure,
   };
 
- protected:
```
Comment (dpallotti) [inline src/CaptureServiceBase/include/CaptureServiceBase/CaptureServiceBase.h:36]: Why this change? Isn't `InitializeCapture` still only used by the subclasses?

File: src/LinuxCaptureService/LinuxCaptureService.cpp [hunk 5]
```diff
@@ -19,22 +19,32 @@ grpc::Status LinuxCaptureService::Capture(
...
     case CaptureInitializationResult::kAlreadyInProgress:
       return {grpc::StatusCode::ALREADY_EXISTS,
               "Cannot start capture because another capture is already in progress"};
   }
 
-  ORBIT_UNREACHABLE();
+  std::shared_ptr<orbit_capture_service_base::StartStopCaptureRequestWaiter>
+      start_stop_capture_request_waiter =
+          orbit_capture_service_base::CreateGrpcStartStopCaptureRequestWaiter(reader_writer);
+  auto capture_options_or_error = start_stop_capture_request_waiter->WaitForStartCaptureRequest();
```
Comment (dpallotti) [inline src/LinuxCaptureService/LinuxCaptureService.cpp:35]: Maybe I'm just missing some context on the cloud collector but something is strange here.
The point of `StartStopCaptureRequestWaiter` was that it's passed to `DoCapture`, method shared by `LinuxCaptureService` and the cloud collector. But now `WaitForStartCaptureRequest` is called outside of it, and `DoCapture` only calls `WaitForStopCaptureRequest`.
So is it even necessary now to have `WaitForStartCaptureRequest` in the interface? Doesn't the shared code between `LinuxCaptureService` and the cloud collector only use `WaitForStopCaptureRequest`? Now where `WaitForStartCaptureRequest` is used, the concrete type is known.

File: src/CaptureServiceBase/include/CaptureServiceBase/CaptureServiceBase.h (hunk 4 above)
Comment (vickyliu-go4it) [reply src/CaptureServiceBase/include/CaptureServiceBase/CaptureServiceBase.h:36]: Previously we just need to call `DoCapture` from the CloudCollector. But now we move the `InitializeCapture` out from `DoCapture` and we will implement the similar logic in `LinuxCaptureService::Capture` for the CloudCollector.

File: src/LinuxCaptureService/LinuxCaptureService.cpp (hunk 5 above)
Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureService.cpp:35]: The `CloudCollector::StartCaptureService` method (this can be found in code search) will implement the similar logic as the `LinuxCaptureService::Capture` and `WindowsCaptureService::Capture` methods. And hence we still use `WaitForStartCaptureRequest` in the CloudCollector.

Comment (dpallotti) [reply src/LinuxCaptureService/LinuxCaptureService.cpp:35]: You still need the logic implemented in `WaitForStartCaptureRequest` by the implementations of `StartStopCaptureRequestWaiter`. What I'm saying is that you don't need that method to be part of the interface anymore, because it's not used in `DoCapture` (which doesn't know about the concrete types), but it's directly used in `LinuxCaptureService::Capture`, `WindowsCaptureService::Capture` and `CloudCollector::StartCaptureService`. All these methods use `WaitForStartCaptureRequest` already knowing the type of the `StartStopCaptureRequestWaiter`. For example, the content of `GrpcStartStopCaptureRequestWaiter::WaitForStartCaptureRequest` can be directly moved here. So the interface can become `StopCaptureRequestWaiter` and contain only `WaitForStopCaptureRequest` as that's the only method used under `DoCapture`.

Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureService.cpp:35]: Got it! Thanks for the explaination!

Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureService.cpp]: I removed `WaitForStartCaptureRequest` method from the interface. The suggestion is really great and now we can have different return type for `WaitForStartCaptureRequest`. Thanks again!

**Review by dpallotti — APPROVED** [review]

**Review by beckerhe — APPROVED** [review]
Sorry for being late. LGTM
